### PR TITLE
rustPlatform.importCargoLock: add support for git dependencies that use workspace inheritance

### PR DESCRIPTION
**Pull request by winterqt**
###### Description of changes

Rust 1.64.0 added support for workspace inheritance, which allows
for crates to inherit values such as dependency version constraints or
package metadata information from their workspaces [0].

This works by having workspace members specify a value as a table, with
`workspace` set to true. Thus, supporting this in importCargoLock is as
simple as walking the crate's Cargo.toml, replacing inherited values
with their workspace counterpart.

This is also what a forthcoming Cargo release will do for `cargo vendor` [1],
but we can get ahead of it ;)

[0]: https://blog.rust-lang.org/2022/09/22/Rust-1.64.0.html#cargo-improvements-workspace-inheritance-and-multi-target-builds
[1]: https://github.com/rust-lang/cargo/pull/11414

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by winterqt** [comment]
I wanted to include bumping Ruff here, but ran into https://github.com/charliermarsh/ruff/issues/3039. If it gets fixed before this is merged, I'll add it.

File: pkgs/build-support/rust/test/import-cargo-lock/git-dependency-workspace-inheritance/default.nix
Comment (figsoda) [inline pkgs/build-support/rust/test/import-cargo-lock/git-dependency-workspace-inheritance/default.nix:1]: instead of testing the behavior of `replaceWorskpaceValues`, what if we just test `cargoLock.lockFile` directly?

```toml
[dependencies.rustpython-compiler-core]
package = "rustpython-compiler-core"
git = "https://github.com/RustPython/RustPython"
```

Comment (winterqt) [reply pkgs/build-support/rust/test/import-cargo-lock/git-dependency-workspace-inheritance/default.nix:1]: We could do that, but I also kind of prefer the simplistic nature of the current test, as it's tiny + adequately tests what the tool does.

Either works -- let's wait for others to weigh in.

File: pkgs/build-support/rust/replace-workspace-values.py
Comment (figsoda) [inline pkgs/build-support/rust/replace-workspace-values.py:21]: > [`features`](https://doc.rust-lang.org/cargo/reference/features.html) declared in this table are additive with the `features` from `[dependencies]`

We should probably handle this

https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table

Comment (winterqt) [reply pkgs/build-support/rust/replace-workspace-values.py:23]: Agreed, done.

File: pkgs/build-support/rust/replace-workspace-values.py
Comment (figsoda) [inline pkgs/build-support/rust/replace-workspace-values.py:31]: I think this should do the same thing, and we can move this logic (and whatever `dict.fromkeys` does I'm assuming) after (or before if we mutate workspace_copy) `replaced.update(workspace_copy)`
```suggestion
            merged_features = set((crate_features or []) + (workspace_copy.get("features") or []))
```

Comment (winterqt) [reply pkgs/build-support/rust/replace-workspace-values.py]: You know, when writing that, I was thinking to myself "if only Python had `or`." 😄

File: pkgs/build-support/rust/replace-workspace-values.py
Comment (figsoda) [inline pkgs/build-support/rust/replace-workspace-values.py:37]:
```suggestion
            for k in root[key].keys():
```

Comment (winterqt) [reply pkgs/build-support/rust/replace-workspace-values.py:39]: Oops, result of a refactor and I didn't change that line. Surprised flake8 didn't warn about it...

File: pkgs/build-support/rust/replace-workspace-values.py
Comment (figsoda) [inline pkgs/build-support/rust/replace-workspace-values.py:65]: is this really worth it?

Comment (figsoda) [reply pkgs/build-support/rust/replace-workspace-values.py:65]: As an alternative, we should probably have a global that is set to true [here](https://github.com/NixOS/nixpkgs/blob/0cc4830f14578ece6a3e0c1e543785570e26f2e4/pkgs/build-support/rust/replace-workspace-values.py#L23), unconditionally `deepcopy`ing the whole table doesn't look like optimization to me, if that's what you were going for

Comment (winterqt) [reply pkgs/build-support/rust/replace-workspace-values.py:65]: Fair, not sure what I was thinking there 😕

Comment (winterqt) [reply pkgs/build-support/rust/replace-workspace-values.py]: (I ended up not using a global, but a bool returned from each function and ORed.)

**Review by figsoda — APPROVED** [review]
Can't speak much about the changes to `matrix-conduit`, would appreciate a review from one of the maintainers.

The `importCargoLock` and `ruff` changes lgtm, would prefer if tests are done using `buildRustPackage` instead but this is up for discussion

**Comment by figsoda** [comment]
Result of `nixpkgs-review pr 217232` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>flare-floss</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>matrix-conduit</li>
    <li>python310Packages.python-flirt</li>
    <li>python311Packages.python-flirt</li>
    <li>ruff</li>
    <li>sapling</li>
  </ul>
</details>

**Comment by winterqt** [comment]
All of those failures are unrelated.

Darwin: dependency being missed after the libiconv changes (I presume, haven't looked at history to confirm).
Linux (and I presume Darwin, after fixing the above): dependency version mismatch?

**Comment by figsoda** [comment]
Yep I was just posting this so other people don't have to rebuild these

**Comment by winterqt** [comment]
Fixing these now...

**Comment by wischli** [comment]
Could you make an estimation when a consumer would be able to make use of this fix inside any nix (pre) release?

**Comment by figsoda** [comment]
This is already in master and the unstable branches, if you are using stable, then 23.05

**Comment by wischli** [comment]
> This is already in master and the unstable branches, if you are using stable, then 23.05

Thanks for the quick reply. Any chance it will be backported to 22.11?

**Comment by figsoda** [comment]
@wischli can you test #223804?

**Comment by wischli** [comment]
> @wischli can you test #223804?

Unfortunately, the error still persists

```
       > Caused by:
       >   error inheriting `version` from workspace root manifest's `workspace.package.version`
```

This is the relevant feature where we encounter this: https://github.com/centrifuge/centrifuge-chain/pull/1241
One of our dependencies ([Polkadot](https://github.com/paritytech/polkadot)) switched to workspace inheritance.

**Comment by figsoda** [comment]
This fix only applies to `importCargoLock`, the fix for `fetchCargoTarball` will come with the rust 1.68 update, which will not be backported to 22.11

```patch
-             cargoSha256 = "sha256-BkSVd08QksOvyIRO1bVpiOnWf440r+dLR1A3UiND7IM=";
+             cargoLock = {
+               lockFile = ./Cargo.lock;
+               allowBuiltinFetchGit = true;
+             };
```

**Comment by wischli** [comment]
Thanks for sharing this information with us. Looking forward to the release which supports workspace inheritance in `fetchCargoTarball`.

**Pull request merged**
